File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/utils/errors/grammar-parse-error.ts
import { buildErrorPosition } from "./build-error-position.js";

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

export class GrammarParseError extends Error {
  override name = "GrammarParseError";
  grammar: string;
  pos: number;
  reason: string;

  constructor(grammar: string, pos: number, reason: string) {
    super(
      [
        GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
        "",
        ...buildErrorPosition(grammar, pos),
      ].join("\n"),
    );
    this.grammar = grammar;
    this.pos = pos;
    this.reason = reason;
  }

  // Mirrors Python's `str(error)`, which returns the message on its own.
  override toString(): string {
    return [
      GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(this.reason),
      "",
      ...buildErrorPosition(this.grammar, this.pos),
    ].join("\n");
  }

  equals(other: unknown): boolean {
    return this.toString() === String(other);
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/utils/errors/grammar_parse_error.py
from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f"Failed to parse grammar: {reason}"


class GrammarParseError(Exception):
    def __init__(self, grammar: str, pos: int, reason: str) -> None:
        super().__init__(
            "\n".join(
                [
                    GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
                    "",
                    *build_error_position(grammar, pos),
                ]
            )
        )
        self.grammar = grammar
        self.pos = pos
        self.reason = reason

    def __str__(self) -> str:
        return "\n".join(
            [
                GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(self.reason),
                "",
                *build_error_position(self.grammar, self.pos),
            ]
        )

    def __eq__(self, other: object) -> bool:
        return str(self) == str(other)

    def __hash__(self) -> int:
        return hash(str(self))
